traits/trait_list_object.py

```diff
@@ -53,460 +94,754 @@ def __init__(self, index=0, removed=None, added=None):
+            elif count <= 0:
+                self.clear()
+        else:
+            raise TypeError(
+                "{} object cannot be interpreted as an integer".format(
+                    type(count)))
+        return self
+
+    # ------------------------------------------------------------------------
+    # Private interface
+    # ------------------------------------------------------------------------
+
+    def _get_removed(self, index, default=[]):
+        """ Compute removed values given index. """
+        try:
+            if not isinstance(index, slice):
+                return [self[index]]
+            else:
+                return self[index]
+        except IndexError:
+            return default
 
-            return result
+    def _normalize(self, index):
+        if isinstance(index, slice):
+            return self._normalize_slice(index)
+        else:
+            return self._normalize_index(index)
 
+    def _normalize_index(self, index):
+        """ Normalize integer index to range 0 to len (inclusive). """
+        index = operator.index(index)
+        if index < 0:
+            return max(0, len(self) + index)
         else:
-            self.len_error(len(self) - 1)
+            return min(len(self), index)
 
-    def rename(self, name):
-        trait = self.object()._trait(name, 0)
-        if trait is not None:
-            self.name = name
-            self.trait = trait.handler
+    def _normalize_slice(self, index):
+        """ Normalize slice start and stop to range 0 to len (inclusive). """
 
-    def clear(self):
-        if self.trait.minlen > 0:
-            self.len_error(0)
+        # Do not normalize if step is negative.
+        if index.step is not None and index.step < 0:
+            return index
 
-        if len(self) == 0:
-            list.clear(self)
-            return
+        if index.step is None or index.step > 0:
+            if index.start is not None:
+                start = self._normalize_index(index.start)
+            else:
+                start = 0
+            if index.stop is not None:
+                stop = self._normalize_index(index.stop)
+            else:
+                stop = len(self)
+        else:
+            if index.start is not None:
+                start = self._normalize_index(index.start)
+            else:
+                start = len(self)
+            if index.stop is not None:
+                stop = self._normalize_index(index.stop)
+            else:
+                stop = 0
 
-        removed = self.copy()
-        list.clear(self)
+        return slice(start, stop, index.step)
 
-        if self.name_items is not None:
-            self._send_trait_items_event(
-                self.name_items, TraitListEvent(0, removed)
-            )
+    def _should_notify(self, removed, added):
+        try:
+            if added == removed:
+                return False
+            return True
+        except Exception:
```

Comment:
Under what circumstances is it necessary to catch an exception here? `except Exception` is very broad, and risks hiding coding errors. Can we cut this down to something more targeted, and add a comment explaining the necessity?

Reply:
No longer used.

Reply:
Comparing two arbitrary objects for equality can raise I assume?
Consider this test, which passes on master but fails on this branch:
```

class MyClass(HasTraits):
    x = List()


class BadClass:

    def __eq__(self, other):
        raise TypeError("Oops")


class TestFooList(unittest.TestCase):
    def test_notify_unvalidated_list(self):
        foo = MyClass()
        foo.x = [BadClass()]

        foo.x[0] = 2.0
        self.assertEqual(foo.x, [2.0])

```

Reply:
In the normal scheme of things, it should be very rare for an `==` or `!=` comparison to raise: that's usually a sign of badly written code somewhere. (Notice that in core Python, pretty much _anything_ can be compared by equality with anything else.)

So unless there's a really pressing practical need to deal with that case, I don't think we should do so.

The one exception to the general rule, and the one possible practical case we _might_ have to deal with, is NumPy arrays.

Reply:
In the case that @kitchoi shows, a `TypeError` from an `__eq__` method would likely be a coding mistake, and I'd expect and want that error to propagate.

Reply:
Understood. Just another change to the current behaviour I suppose. In any case, I will kill this equality comparison for EEP3.

Reply:
> The one exception to the general rule, and the one possible practical case we might have to deal with, is NumPy arrays.

E.g.:
```
class MyClass(HasTraits):
    x = List()

import numpy

class TestFooList(unittest.TestCase):
    def test_notify_unvalidated_list(self):
        foo = MyClass()
        foo.x = [numpy.array([1.0])]
        foo.x[0] = numpy.array([2.0, 3.0])

```

```
Traceback (most recent call last):
  File "/Users/kchoi/Work/ETS/traits/traits/tests/test_list.py", line 388, in test_notify_unvalidated_list
    foo.x[0] = numpy.array([2.0, 3.0])
  File "/Users/kchoi/Work/ETS/traits/traits/trait_list_object.py", line 291, in __setitem__
    if added != removed:
ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
```

Reply:
Yes, exactly. Honestly, I'm not sure whether it's worth adding special-case code for this use-case.

Reply:
If we didn't have any backwards compatibility constraints, the behaviour I'd want to see here is that `added` and `removed` aren't compared at all - instead, the event simply reports the added and removed items, even if they happen to be the same.

Reply:
I agree that we should not add special-case code for this.
This `if added != removed` is for silencing notifications if the changed items compared equally.  In my opinion, it is trying to be too clever and is guessing what the user wants.
We will need to remove this for EEP 3 anyway. Just before it is removed and after this PR is merged, mutating trait lists with numpy arrays may produce the error shown above.

Reply:
> In my opinion, it is trying to be too clever and is guessing what the user wants.

Absolutely agreed.